atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java

```diff
@@ -68,18 +66,102 @@ public SchemaMutationLock(
         void execute() throws Exception;
     }
 
-    public void runWithLock(Action action) {
-        long lockId = waitForSchemaMutationLock();
+    private class HeartbeatThread {
+        private final Thread heartbeatThread;
+        private final long lockId;
+        private volatile boolean keepBeating;
+        private volatile long heartbeatCount;
+
+        HeartbeatThread(long lockId) {
+            this.heartbeatThread = new Thread(this::heartbeat);
+            this.lockId = lockId;
+            this.heartbeatCount = 0;
+        }
+
+        void startBeating() {
+            keepBeating = true;
+            heartbeatThread.start();
+        }
+
+        void stopBeating() {
+            keepBeating = false;
+            try {
+                // make sure beating's stopped
+                heartbeatThread.join();
+            } catch (InterruptedException e) {
+                LOGGER.debug("Interrupted while trying to stop the heartbeat heartbeatThread.");
+            }
+        }
+
+        void heartbeat() {
+            while (keepBeating) {
+                try {
+                    clientPool.runWithRetry((FunctionCheckedException<Cassandra.Client, Void, Exception>) client -> {
+                        Column ourUpdate = lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount + 1));
+
+                        List<Column> expected = ImmutableList.of(lockColumnWithStrValue(String.format(
+                                CassandraConstants.GLOBAL_DDL_LOCK_FORMAT, lockId, heartbeatCount)));
+
+                        CASResult casResult = writeDdlLockWithCas(client, expected, ourUpdate);
+
+                        if (casResult.isSuccess()) {
+                            heartbeatCount++;
+                        } else {
+                            handleForcedLockClear(casResult, lockId, heartbeatCount);
+                        }
+                        return null;
+                    });
+                    Thread.sleep(configManager.getConfig().heartbeatTimePeriodMillis());
+                } catch (Exception e) {
```

Comment:
I.e. We probably want to test cancellation and determine if we should be resetting the current thread's interrupt flag.